### ISSUE TITLE
Feature proposal: Differentiable AUC approximation loss function for binary classification

### ISSUE THREAD

**Issue by taromakino** [Lasagne/Lasagne#767]
The area under the ROC curve (AUC) metric is non-differentiable, but a differentiable approximation exists (Calders & Jaroszewicz, 2007) that is based on the AUC's interpretation as the probability that the classifier ranks a randomly chosen true positive higher than a randomly chosen false positive (H.L. van Trees, 1968).

I have implemented the function on the bottom of the second page of Calders & Jaroszewicz (2007) using Theano functions, and have verified that the output values match those of sklearn.metrics.roc_auc_score for large enough sample sizes.  Assuming I correctly follow the contribution guidelines, would a pull request with this feature be accepted?  The AUROC is a common evaluation metric for binary classification competitions on Kaggle, and I noticed some demand for it when searching the internet for an implementation.  Also, I'm not sure whether this would belong in lasagne or in theano.tensor.nnet, so I will be posting on there as well.

**Comment by f0k** [comment]
> Assuming I correctly follow the contribution guidelines, would a pull request with this feature be accepted?

In general, this sounds like a useful addition to me, even if probably a bit exotic. I see that it has been used e.g. in https://arxiv.org/abs/1511.05265. Are you aware of other uses? How experimental is this? Also, how complex is the code? Feel free to post the cost function here or as a gist. If it's not extremely complex and it works, we should probably include it. If it's dangerously experimental, it might be suited better as a Lasagne Recipe.

> the output values match those of sklearn.metrics.roc_auc_score for large enough sample sizes

I haven't read the papers yet -- does it still do something useful when training on small mini-batches? If it's available in `lasagne.objectives`, people will try to swap it in for other losses, and we should at least document well under what circumstances it will work.

**Comment by taromakino** [comment]
I have been using it with batch size of 256, and it has been effective in maximizing the AUROC over the entire training set.  The code is simple, I use dimshuffle and broadcasting to take the difference of every pair of positive and negative predictions.

```
import theano

def calc_auroc_loss(pred_vr, y_vr):
    pos_pred_vr = pred_vr[y_vr.nonzero()]
    neg_pred_vr = pred_vr[theano.tensor.eq(y_vr, 0).nonzero()]
    pred_diffs_vr = pos_pred_vr.dimshuffle(0, 'x') - neg_pred_vr.dimshuffle('x', 0)
    pred_diffs_vr = theano.tensor.ge(pred_diffs_vr, 0)
    num_pairs_vr = theano.tensor.sum(theano.tensor.eq(y_vr, 1)) * theano.tensor.sum(theano.tensor.eq(y_vr, 0),
        dtype='float32')
    auroc_vr = theano.tensor.sum(pred_diffs_vr) / num_pairs_vr
    return -auroc_vr
```

**Comment by f0k** [comment]
> I have been the version below with batch size of 256

The problem is that it'll break when all examples in a batch happen to be of the same class. Other means of balancing positive and negative examples use some precomputed statistics (like the total proportion of positive examples) to become independent from the batches.

> this version, called Soft-AUC in Calders & Jaroszewicz (2007)

Could you link the paper here?

> Here is my code for the Soft-AUC, which uses dimshuffle and broadcasting to sum over each pair of positive and negative predictions.

Looks legit, except that I'd use `theano.tensor.mean` for `auroc_vr` so you don't need to compute `num_pairs_vr`. And I believe `~y_vr.nonzero()` would work instead of `theano.tensor.eq(y_vr, 0).nonzero()`. A downside is that this formulation does not work for multiple binary classifications (unlike binary_crossentropy), and that it returns a scalar loss (unlike all the others in `lasagne.objectives`, which return an elementwise or itemwise loss).

> does not return values similar to sklearn.metrics.roc_auc_score

That's because you use a sigmoid instead of a Heaviside step, right?

I'm hesitating because it's not an established loss function. @benanne, what do you think?

**Comment by taromakino** [comment]
Thank you for the code edits @f0k, I agree with not taking the mean manually.  However, `~y_vr.nonzero()` does not seem to work.

I agree with my implementation not working if there are no examples from both classes present in the batch.  What do you think would be an appropriate return value in this case?

I met with the authors of https://arxiv.org/abs/1511.05265 the other day, and learned that they do not use SGD due to an issue with conditional random fields, and are computing the loss on the entire training set every iteration.

**Comment by f0k** [comment]
> However, `~y_vr.nonzero()` does not seem to work.

Sorry, I thought it would give a boolean variable. Boolean variables have only been added recently, though, and boolean indexing is still under way (https://github.com/Theano/Theano/issues/5192). It could be `(1-y_vr).nonzero()` then, but that's probably not going to be any simpler or faster than your version.

> I agree with my implementation not working if there are no examples from both classes present in the batch. What do you think would be an appropriate return value in this case?

Probably zero, to just skip that batch without doing any harm (although this would still affect SGD+ methods like momentum or adam). Getting it to work nicely, without discarding batches, may need a clever batch iteration scheme or a pool of positive and negative examples kept around to augment batches as needed. But at that point it's not a drop-in replacement for existing loss functions.

> computing the loss on the entire training set every iteration

This circumvents many problems, but opens others.

Given that this is not an established loss and getting it to play nicely with stochastic updates might require some research, I guess we shouldn't include it into Lasagne for now and leave it up to users to copy/paste it if they want to play with it. If this loss can be shown to give an advantage over conventional losses (in a research paper or even just in blog posts or Kaggle competitions), we can include it with corresponding references. I highly encourage you to experiment with it and would be interested to hear back!

I'm closing this for now. @benanne, please reopen if you disagree!
